metric_learn/base_metric.py

```diff
@@ -396,6 +395,150 @@ def _set_default_threshold(self, pairs, y):
+          threshold
+
+    beta : float in [0, 1], optional (default=None)
+      beta value to be used in case strategy == 'f_beta'
+
+    threshold : float in [0, 1] or None, (default=None)
+      In case strategy is 'max_tpr' or 'max_tnr' this parameter must be set
+      to specify the threshold for the true negative rate or true positive
+      rate respectively that needs to be achieved
+
+    pairs_valid : array-like, shape=(n_pairs_valid, 2, n_features)
+      The validation set of pairs to use to set the threshold.
+
+    y_valid : array-like, shape=(n_pairs_valid,)
+      The labels of the pairs of the validation set to use to set the
+      threshold.
+
+    References
+    ----------
+    .. [1] Receiver-operating characteristic (ROC) plots: a fundamental
+           evaluation tool in clinical medicine, MH Zweig, G Campbell -
+           Clinical chemistry, 1993
+
+    .. [2] most of the code of this function is from scikit-learn's PR #10117
+
+    See Also
+    --------
+    sklearn.calibration : scikit-learn's module for calibrating classifiers
+    """
+
+    if strategy not in ('accuracy', 'f_beta', 'max_tpr',
+                        'max_tnr'):
+      raise ValueError('Strategy can either be "accuracy", "f_beta" or '
+                       '"max_tpr" or "max_tnr". Got "{}" instead.'
+                       .format(strategy))
+
+    if strategy == 'max_tpr' or strategy == 'max_tnr':
+      if (threshold is None or not isinstance(threshold, (int, float)) or
+          not threshold >= 0 or not threshold <= 1):
+        raise ValueError('Parameter threshold must be a number in'
+                         '[0, 1]. '
+                         'Got {} instead.'.format(threshold))
+
+    if strategy == 'f_beta':
+      if beta is None or not isinstance(beta, (int, float)):
+        raise ValueError('Parameter beta must be a real number. '
+                         'Got {} instead.'.format(type(beta)))
+
+    pairs_valid, y_valid = self._prepare_inputs(pairs_valid, y_valid,
+                                                type_of_inputs='tuples')
+
+    n_samples = pairs_valid.shape[0]
+    if strategy == 'accuracy':
+      scores = self.decision_function(pairs_valid)
+      scores_sorted_idces = np.argsort(scores)[::-1]
+      scores_sorted = scores[scores_sorted_idces]
+      # true labels ordered by decision_function value: (higher first)
+      y_ordered = y_valid[scores_sorted_idces]
+      # finds the threshold that maximizes the accuracy:
+      cum_tp = stable_cumsum(y_ordered == 1)  # cumulative number of true
+      # positives
+      cum_tn_inverted = stable_cumsum(y_ordered[::-1] == -1)
+      cum_tn = np.concatenate([[0], cum_tn_inverted[:-1]])[::-1]
+      cum_accuracy = (cum_tp + cum_tn) / n_samples
+      max_i = np.argmax(cum_accuracy)
+      # note: we want a positive threshold (distance), so we take - threshold
+      self.threshold_ = - scores_sorted[max_i]
```

Comment:
For the threshold, I use a threshold which is taken among the values of the train set scores (i.e. some/one of the train set scores value(s) on the trainset will be equal to the threshold). But I thought it could also make sense to set a threshold that is  the mean between the value of the lowest value of the train set scores that we want to keep and the highest value of the train set scores that we want to reject, doing sort of a margin for classification. Would that be a better strategy ? Is there other srategies that are better ?

Reply:
In principle what you suggest is a good idea. Not sure what they do in the sklearn PR?

Reply:
I guess that the same reasoning could be applied when setting the threshold for other calibration strategies (other than accuracy)?

Reply:
Yes, I would use this strategy for all strategies

I was about to implement it but then I wondered: what would we do in case the best threshold is at an extreme point (ex: we want to take all points)? Would we say that the optimal threshold is the score of this point (i.e. the worst score) ? Or the score of this point - 1 ? I know this case would rarely happen, but we should need to decide what to do... I think we could go for just the score of the point, it seems less arbitrary...

Note: scikit-learn can use the +1 strategy see https://github.com/scikit-learn/scikit-learn/blob/62d205980446a1abc1065f4332fd74eee57fcf73/sklearn/metrics/ranking.py#L642 (but here it's more justified, it's because we want to have a case where we reject all points. Since the equality is tested with the threshold when classifying, it's normal to have something higher than all scores)
By the way I need to deal with this case: I've added it to the TODO